src/lib/utils/DY.Lib.SplitFunction.fst

```diff
@@ -0,0 +1,283 @@
+  Lemma
+  (requires
+    all_disjoint (func.is_disjoint) (List.Tot.map fst lfuns) /\
+    List.Tot.memP (tag_set, lfun) lfuns
+  )
+  (ensures has_local_fun func (mk_global_fun func lfuns) (tag_set, lfun))
+let mk_global_fun_correct func lfuns tag_set lfun =
+  reveal_opaque (`%mk_global_fun) (mk_global_fun);
+  introduce
+    forall tagged_data.
+      match func.decode_tagged_data tagged_data with
+      | Some (tag, raw_data) ->
+        tag `func.tag_belong_to` tag_set ==> (func.apply_global_fun (mk_global_fun func lfuns) tagged_data == func.apply_local_fun lfun raw_data)
+      | None -> True
+  with (
+    match func.decode_tagged_data tagged_data with
+    | Some (tag, raw_data) -> (
+      if tag `func.tag_belong_to` tag_set then (
+        mk_global_fun_correct_aux func lfuns tag_set lfun tag;
+        func.apply_mk_global_fun (mk_global_fun_aux func lfuns) tagged_data
+      ) else ()
+    )
+    | None -> ()
+  )
+
+/// Equivalence theorem for `mk_global_fun`.
+/// This may be useful to lift properties of the local functions
+/// to property on the global function.
+/// (e.g. the function keep the same output when the trace grows.)
+
+val mk_global_fun_eq:
+  func:split_function_input_values -> lfuns:list (func.tag_set_t & func.local_fun) ->
+  tagged_data:func.tagged_data_t ->
+  Lemma (
+    func.apply_global_fun (mk_global_fun func lfuns) tagged_data == (
+      match func.decode_tagged_data tagged_data with
+      | Some (tag, raw_data) -> (
+        match find_local_fun func lfuns tag with
+        | Some lfun -> func.apply_local_fun lfun raw_data
+        | None -> func.apply_global_fun func.default_global_fun tagged_data
+      )
+      | None -> func.apply_global_fun func.default_global_fun tagged_data
+    )
+  )
+let mk_global_fun_eq func lfuns tagged_data =
+  reveal_opaque (`%mk_global_fun) (mk_global_fun);
+  func.apply_mk_global_fun (mk_global_fun_aux func lfuns) tagged_data
+
+/// If a global function contains some local function,
+/// and the global function input decodes to a tag for this local function,
+/// then the global function is equivalent to the local function on this input.
+
+val local_eq_global_lemma:
+  func:split_function_input_values ->
+  gfun:func.global_fun -> tag_set:func.tag_set_t -> lfun:func.local_fun ->
+  tagged_data:func.tagged_data_t -> tag:func.tag_t -> raw_data:func.raw_data_t ->
+  Lemma
+  (requires
+    func.decode_tagged_data tagged_data == Some (tag, raw_data) /\
+    tag `func.tag_belong_to` tag_set /\
+    has_local_fun func gfun (tag_set, lfun)
+  )
+  (ensures func.apply_global_fun gfun tagged_data == func.apply_local_fun lfun raw_data)
+let local_eq_global_lemma func gfun tag_set lfun tagged_data tag raw_data = ()
+
+/// In the common case where tag disjointness is unequality,
+/// `all_disjoint` is equivalent to the `no_repeats_p` predicate
+/// from F*'s standard library.
+
+val default_disjoint:
```

Comment:
i'd prefer if this name includes "inequality" or "unequal", so that when we use this function (for example in Events or State.Tagged) one can immediately see what it means.
I can think of several options "default_disjoint_inequality", "disjoint_inequality", "inequality_disjoint", ...
Or maybe even just "unequal"? Then it would read nicely in, say, Event: `is_disjoint = unequal;`

Reply:
I went for `unequal`, this is a better name indeed